src/rust/collections/concurrent_ring.rs

```diff
@@ -0,0 +1,509 @@
+            copy(pop_ptr, buf_ptr, pop_len - HEADER_SIZE)
+        };
+
+        // Move to next buffer.
+        self.free_space(pop_offset, pop_len);
+        Ok(pop_len - HEADER_SIZE)
+    }
+
+    /// Removes the next message from the ring buffer up to [len] bytes and copies into [buf]. This function may block
+    /// (spin).
+    #[allow(unused)]
+    pub fn pop(&self, buf: &mut [u8]) -> usize {
+        loop {
+            if let Ok(pop_len) = self.try_pop(buf) {
+                break pop_len;
+            }
+        }
+    }
+
+    /// Atomically writes a header at the indicated offset and returns the previous one.
+    fn write_header(&self, offset: usize, val: usize) -> usize {
+        let buffer_ptr: *mut u8 = unsafe { self.buffer.get_mut() }.as_mut_ptr();
+        let header_ptr: *mut u16 = unsafe { buffer_ptr.add(offset) } as *mut u16;
+        let header: &AtomicU16 = AtomicU16::from_mut(unsafe { &mut *header_ptr });
+        header.swap(val as u16, atomic::Ordering::Relaxed) as usize
+    }
+
+    /// Given a [push_offset] and [pop_offset] into the ring buffer, return available space for writing data. Always
+    /// leave one [HEADER_SIZE] space for distinguishing a full from empty buffer.
+    fn available_space(&self, push_offset: usize, pop_offset: usize) -> usize {
+        let used_space: usize = (push_offset + self.capacity() - pop_offset) % self.capacity();
+        self.capacity() - used_space - HEADER_SIZE
+    }
+
+    /// Align to u16 for the header offset.
+    fn align_header(offset: usize) -> usize {
+        // Round up to u16 aligned.
+        match offset % HEADER_SIZE {
+            0 => offset,
+            x => offset - x + HEADER_SIZE,
+        }
+    }
+
+    /// Reserves [len] bytes from the ring buffer. If successful, returns the offset of the beginning of the buffer,
+    /// else returns None.
+    fn allocate_space(&self, len: usize) -> Option<usize> {
+        let len_: usize = Self::align_header(len);
+        let push_offset: usize = peek(self.push_offset);
+        let pop_offset: usize = peek(self.pop_offset);
+
+        if len_ > self.available_space(push_offset, pop_offset) {
+            return None;
+        }
+        let new_offset: usize = (push_offset + len_) % self.capacity();
+        assert_ne!(new_offset, pop_offset);
+        // Queue has space after the enqueue pointer, so try to reserve space.
+        match check_and_set(self.push_offset, push_offset, new_offset) {
+            Ok(start) => Some(start),
+            Err(_) => None,
+        }
+    }
+
+    /// Frees [len] bytes from the ring buffer.
+    fn free_space(&self, current_offset: usize, len: usize) {
+        let len_: usize = Self::align_header(len);
+        let new_offset: usize = current_offset + len_ % self.capacity();
+        // Ensure that the old pop_offset was what we expected.
+        let old: usize = match check_and_set(self.pop_offset, current_offset, new_offset) {
+            Ok(old) => old,
```

Comment:
If we always need the old value, do we need a match here?

Reply:
This is fixed.